test/model/message_test.dart

```diff
@@ -473,36 +634,220 @@ void main() {
+      await prepare();
+      await store.handleEvent(eg.submessageEvent(1000, eg.selfUser.userId,
+        content: PollQuestionEventSubmessage(question: 'New question')));
       checkNotNotified();
-      check(store).messages.values.single.id.equals(message1.id);
     });
 
-    test('delete messages', () async {
-      final message1 = eg.streamMessage();
-      final message2 = eg.streamMessage();
+    test('message has no submessages', () async {
+      final message = eg.streamMessage();
       await prepare();
-      await prepareMessages([message1, message2]);
-      await store.handleEvent(eg.deleteMessageEvent([message1, message2]));
-      checkNotifiedOnce();
-      check(store).messages.isEmpty();
+      await prepareMessages([message]);
+      await store.handleEvent(eg.submessageEvent(message.id, eg.otherUser.userId,
+        content: PollQuestionEventSubmessage(question: 'New question')));
+      checkNotNotified();
+      check(store.messages[message.id]).isNotNull().poll.isNull();
     });
 
-    test('delete an unknown message with a known message', () async {
-      final message1 = eg.streamMessage();
-      final message2 = eg.streamMessage();
-      final message3 = eg.streamMessage();
-      await prepare();
-      await prepareMessages([message1, message2]);
-      await store.handleEvent(eg.deleteMessageEvent([message2, message3]));
-      checkNotifiedOnce();
-      check(store).messages.values.single.id.equals(message1.id);
+    group('question event', () {
+      test('update question', () async {
+        final message = await preparePollMessage(question: 'Old question');
+        await store.handleEvent(eg.submessageEvent(message.id, eg.selfUser.userId,
+          content: PollQuestionEventSubmessage(question: 'New question')));
+        checkNotifiedOnce();
+        checkPoll(message).question.equals('New question');
+      });
+
+      test('unauthorized question edits', () async {
+        final message = await preparePollMessage(
+          question: 'Old question',
+          messageSender: eg.otherUser,
+        );
+        checkPoll(message).question.equals('Old question');
+        await store.handleEvent(eg.submessageEvent(message.id, eg.selfUser.userId,
+          content: PollQuestionEventSubmessage(question: 'edit')));
+        checkPoll(message).question.equals('Old question');
+      });
+
+      test('ignore submessage event with malformed content', () async {
+        final message = await preparePollMessage(question: 'Old question');
+        await checkLogs(() => store.handleEvent(SubmessageEvent(
+          id: 0, msgType: SubmessageType.widget, submessageId: 123,
+          messageId: message.id,
+          senderId: eg.selfUser.userId,
+          content: jsonEncode({
+            'type': 'question',
+            // Invalid type for question
+            'question': 100,
+          }))))..length.equals(2)
+               ..last.contains('Malformed submessage event data for poll');
```

Comment:
So, thinking about this `checkLogs` sort of test check now that I've reached this commit:

This feels to me like the sort of thing that's going to be quite brittle. (E.g. why length 2 — I guess the first is the one in `PerAccountStore.handleEvent`?) If we start widely adopting this pattern, then any time we add or remove a `debugLog` call there's going to be a bunch of random tests all over that need to be updated for it.

Also when we change the content of a `debugLog` call, we'll have to update the corresponding test. That's OK for a lot of things (if it's really the test aimed at the thing one is changing), but it makes changes more heavyweight. And these log messages are internal details, in the sense that they're aimed only at us the developers — they're not something our users, or any external system, ever see.

If we find a place where there's a `debugLog` call that we feel is really critical to ensure doesn't accidentally get broken, then that could be a good use case for something like this despite those costs. But I think this `debugLog` call is typical in that it isn't critical to that degree. Let's therefore leave this check out.

Reply:
(As for the server tests:

    This is inspired by `TestCase.assertLogs`, which is widely adapted
    in the Zulip server tests.

I have the sense that a lot of those uses are motivated mostly by the need to *suppress* the logs so they don't spew into the output of running the tests. (That's especially clear when they don't end up asserting anything about the logs emitted; and I suspect it's the only important motivation in many tests that do assert something about them too.) In this codebase we just handle that systematically using `debugLogEnabled`.

The other motivation for `assertLogs` checks in the server is that in the server, those log messages go into the production server log, which is an important operational tool for Zulip server operators including ourselves. That has somewhat the same shape as our `debugLog` calls for debugging… with the crucial difference that when we see `debugLog` output, the app is a debug build, and so we're presumably sitting there debugging it and can easily add more log lines as we wish and then hot-reload and repeat whatever we did. That's very different from the scenario where something went wrong an hour ago on one's live production server and you're frantically trying to pin down what happened.)

Reply:
Thanks for the explanation! Dropping the commit.

Another motivation for adding `checkLogs` was to have a convenient way of verifying we expect to happen in our code -- too convenient that it relies heavily on implementation details. I agree that testing externally visible interfaces are preferred.